test/robot/job/worker/action.go

```diff
@@ -95,12 +95,13 @@ func (a *Actions) apply(ctx context.Context, item Action) error {
 
 func (a *Actions) do(ctx context.Context, w *Worker, input Input) (string, error) {
 	a.mu.Lock()
-	defer a.mu.Unlock()
 	action := proto.Clone(a.nullAction).(Action)
 	action.Init(id.Unique().String(), input, w.Info)
 	if err := a.ledger.Add(ctx, action); err != nil {
 		return "", err
 	}
+	// unlock here so other do calls can be invoked after adding to ledger
```

Comment:
To avoid any problems down the line, it's probably worth creating a small private function that does the clone, init and add while holding the lock, so you can still use the defer to unlock. Using the defer is the preferred pattern, as it ensure that it will *always* unlock - even if another developer later on adds an early return, e.g.